src/main/java/com/edmunds/tools/databricks/maven/UpsertClusterMojo.java

```diff
@@ -142,64 +157,79 @@ private void editCluster(ClusterTemplateModel ct, String clusterId) throws IOExc
-
-        LibraryService libraryService = getDatabricksServiceFactory().getLibraryService();
-        LibraryDTO[] libsToDelete = Arrays.stream(libraryService.clusterStatus(clusterId).getLibraryFullStatuses())
-                // no need to delete shared libraries or libraries we're going to install next
-                .filter(libStatus -> !libStatus.isLibraryForAllClusters() || willBeInstalled.contains(libStatus.getLibrary()))
+    private Set<LibraryDTO> getClusterLibraries(String clusterId) throws IOException, DatabricksRestException {
+        return Arrays.stream(getDatabricksServiceFactory().getLibraryService().clusterStatus(clusterId).getLibraryFullStatuses())
+                .filter(status -> !status.isLibraryForAllClusters())
                 .map(LibraryFullStatusDTO::getLibrary)
-                .toArray(LibraryDTO[]::new);
+                .collect(Collectors.toSet());
+    }
 
-        if (ArrayUtils.isNotEmpty(libsToDelete)) {
-            getLog().info("Libraries to delete: " + Arrays.toString(libsToDelete));
-            libraryService.uninstall(clusterId, libsToDelete);
+    private void detachLibraries(ClusterTemplateModel ct, String clusterId, Set<LibraryDTO> clusterLibraries) throws IOException, DatabricksRestException {
+        getLog().info(String.format("Removing libraries from the cluster: name=[%s], id=[%s]", ct.getClusterName(), clusterId));
+        Set<LibraryDTO> libsToDelete = getLibrariesToDelete(clusterLibraries, ct.getArtifactPaths());
+        if (CollectionUtils.isNotEmpty(libsToDelete)) {
+            LibraryService libraryService = getDatabricksServiceFactory().getLibraryService();
+            libraryService.uninstall(clusterId, libsToDelete.toArray(new LibraryDTO[]{}));
         }
     }
 
-    private void attachLibraries(ClusterTemplateModel ct, String clusterId, LibraryDTO[] libsToInstall) throws IOException, DatabricksRestException {
+    private void attachLibraries(ClusterTemplateModel ct, String clusterId, Set<LibraryDTO> clusterLibraries) throws IOException, DatabricksRestException {
         getLog().info(String.format("Attaching libraries to the cluster: name=[%s], id=[%s]", ct.getClusterName(), clusterId));
-
-        if (ArrayUtils.isNotEmpty(libsToInstall)) {
-            getLog().info("Libraries to install: " + Arrays.toString(libsToInstall));
-            getDatabricksServiceFactory().getLibraryService().install(clusterId, libsToInstall);
+        Set<LibraryDTO> libsToInstall = getLibrariesToInstall(clusterLibraries, ct.getArtifactPaths());
+        if (CollectionUtils.isNotEmpty(libsToInstall)) {
+            getDatabricksServiceFactory().getLibraryService().install(clusterId, libsToInstall.toArray(new LibraryDTO[]{}));
         }
     }
 
-    private LibraryDTO[] getLibrariesToInstall(Collection<String> artifactPaths) {
+    private Set<LibraryDTO> getLibrariesToInstall(Set<LibraryDTO> clusterLibraries, Collection<String> artifactPaths) {
```

Comment:
It looks like this is mostly just filtering on jar libraries, maybe that could be reflected in the method name?

Reply:
Documented this method to clarify its purpose, thank you!